File: tests/cuda/python/test_unary.py
"""Unary CUDA binding tests — aligned with C++ test_unary.cpp.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda
    python -m pytest tests/cuda/python/test_unary.py -v
"""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr, dtype=None):
    if dtype is not None:
        np_arr = np_arr.astype(dtype)
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.to(CPU).numpy()


class TestUnaryCUDA:
    """Unary math operations — CUDA backend."""

    def test_abs(self):
        a_np = np.array([-3, -1, 0, 2, 4], dtype=np.float64)
        result = ins.abs(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.abs(a_np), rtol=1e-10)

    def test_sqrt(self):
        a_np = np.array([1, 4, 9, 16], dtype=np.float64)
        result = ins.sqrt(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.sqrt(a_np), rtol=1e-10)

    def test_exp(self):
        a_np = np.array([0, 1, 2], dtype=np.float64)
        result = ins.exp(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.exp(a_np), rtol=1e-10)

    def test_log(self):
        a_np = np.array([1, 2, 3], dtype=np.float64)
        result = ins.log(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.log(a_np), rtol=1e-10)

    def test_sin_cos_tan(self):
        a_np = np.array([0, 0.5, 1.0], dtype=np.float64)
        np.testing.assert_allclose(to_numpy(ins.sin(to_gpu(a_np))), np.sin(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.cos(to_gpu(a_np))), np.cos(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.tan(to_gpu(a_np))), np.tan(a_np), rtol=1e-10)

    def test_floor_ceil_round(self):
        a_np = np.array([1.2, 2.5, 3.7, -1.3], dtype=np.float64)
        np.testing.assert_allclose(to_numpy(ins.floor(to_gpu(a_np))), np.floor(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.ceil(to_gpu(a_np))), np.ceil(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.round(to_gpu(a_np))), np.round(a_np), rtol=1e-10)

    def test_sign(self):
        a_np = np.array([-3, -1, 0, 2, 4], dtype=np.float64)
        result = ins.sign(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.sign(a_np), rtol=1e-10)

    def test_negative(self):
        a_np = np.array([1, -2, 3], dtype=np.float64)
        result = ins.negative(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), -a_np, rtol=1e-10)

    def test_square(self):
        a_np = np.array([1, 2, 3, 4], dtype=np.float64)
        result = ins.square(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), a_np**2, rtol=1e-10)

    def test_reciprocal(self):
        a_np = np.array([1, 2, 4, 5], dtype=np.float64)
        result = ins.reciprocal(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), 1.0 / a_np, rtol=1e-10)

    def test_log2_log10(self):
        a_np = np.array([1, 2, 8, 100], dtype=np.float64)
        np.testing.assert_allclose(to_numpy(ins.log2(to_gpu(a_np))), np.log2(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.log10(to_gpu(a_np))), np.log10(a_np), rtol=1e-10)

    def test_asin_acos_atan(self):
        a_np = np.array([-0.5, 0, 0.5], dtype=np.float64)
        np.testing.assert_allclose(to_numpy(ins.asin(to_gpu(a_np))), np.arcsin(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.acos(to_gpu(a_np))), np.arccos(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.atan(to_gpu(a_np))), np.arctan(a_np), rtol=1e-10)

    def test_sinh_cosh_tanh(self):
        a_np = np.array([-1, 0, 1], dtype=np.float64)
        np.testing.assert_allclose(to_numpy(ins.sinh(to_gpu(a_np))), np.sinh(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.cosh(to_gpu(a_np))), np.cosh(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.tanh(to_gpu(a_np))), np.tanh(a_np), rtol=1e-10)

    def test_deg2rad_rad2deg(self):
        a_np = np.array([0, 90, 180, 360], dtype=np.float64)
        np.testing.assert_allclose(
            to_numpy(ins.deg2rad(to_gpu(a_np))), np.deg2rad(a_np), rtol=1e-10
        )
        b_np = np.array([0, np.pi / 2, np.pi, 2 * np.pi], dtype=np.float64)
        np.testing.assert_allclose(
            to_numpy(ins.rad2deg(to_gpu(b_np))), np.rad2deg(b_np), rtol=1e-10
        )

    def test_exp2_expm1_log1p(self):
        a_np = np.array([0, 1, 2, 3], dtype=np.float64)
        np.testing.assert_allclose(to_numpy(ins.exp2(to_gpu(a_np))), np.exp2(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.expm1(to_gpu(a_np))), np.expm1(a_np), rtol=1e-10)
        np.testing.assert_allclose(to_numpy(ins.log1p(to_gpu(a_np))), np.log1p(a_np), rtol=1e-10)

    def test_cbrt(self):
        a_np = np.array([1, 8, 27], dtype=np.float64)
        result = ins.cbrt(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.cbrt(a_np), rtol=1e-10)

    def test_trunc(self):
        a_np = np.array([1.2, -2.7, 3.5], dtype=np.float64)
        result = ins.trunc(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.trunc(a_np), rtol=1e-10)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cuda/julia/test_unary.jl
# Unary CUDA tests — aligned with C++ test_unary.cpp
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_unary.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0; failed = 0
function check(name, cond)
    global passed, failed
    if cond; passed += 1; else; failed += 1; println("FAIL: $name"); end
end

println("=== Unary CUDA ===")

u = Insight.to_array(Insight.from_data([-3.0, -1.0, 0.0, 2.0, 4.0]), Insight.GPUPlace(0))

# abs
check("abs", Insight.abs_fn(u).ptr != C_NULL)

# sqrt
v = Insight.to_array(Insight.from_data([1.0, 4.0, 9.0, 16.0]), Insight.GPUPlace(0))
check("sqrt", Insight.sqrt_fn(v).ptr != C_NULL)

# exp
w = Insight.to_array(Insight.from_data([0.0, 1.0, 2.0]), Insight.GPUPlace(0))
check("exp", Insight.exp_fn(w).ptr != C_NULL)

# log
check("log", Insight.log_fn(w).ptr != C_NULL)

# sin
x = Insight.to_array(Insight.from_data([0.0, 0.5, 1.0]), Insight.GPUPlace(0))
check("sin", Insight.sin_fn(x).ptr != C_NULL)

# cos
check("cos", Insight.cos_fn(x).ptr != C_NULL)

# tan
check("tan", Insight.tan_fn(x).ptr != C_NULL)

# floor
y = Insight.to_array(Insight.from_data([1.2, 2.5, 3.7, -1.3]), Insight.GPUPlace(0))
check("floor", Insight.floor_fn(y).ptr != C_NULL)

# ceil
check("ceil", Insight.ceil_fn(y).ptr != C_NULL)

# round
check("round", Insight.round_fn(y).ptr != C_NULL)

# negative
check("negative", Insight.negative(u).ptr != C_NULL)

# exp2
z = Insight.to_array(Insight.linspace(0.1, 2.0, Int64(10), Insight.float32), Insight.GPUPlace(0))
check("exp2", Insight.exp2_fn(z).ptr != C_NULL)

# expm1
check("expm1", Insight.expm1_fn(z).ptr != C_NULL)

# log1p
check("log1p", Insight.log1p_fn(z).ptr != C_NULL)

# cbrt
check("cbrt", Insight.cbrt_fn(z).ptr != C_NULL)

# reciprocal
check("reciprocal", Insight.reciprocal_fn(z).ptr != C_NULL)

# asinh
check("asinh", Insight.asinh_fn(z).ptr != C_NULL)

# acosh
w2 = Insight.to_array(Insight.from_data([1.5, 2.0, 3.0]), Insight.GPUPlace(0))
check("acosh", Insight.acosh_fn(w2).ptr != C_NULL)

# deg2rad
d = Insight.to_array(Insight.from_data([0.0, 90.0, 180.0, 360.0]), Insight.GPUPlace(0))
check("deg2rad", Insight.deg2rad_fn(d).ptr != C_NULL)

# rad2deg
r = Insight.to_array(Insight.from_data([0.0, 1.5707963, 3.1415926, 6.2831853]), Insight.GPUPlace(0))
check("rad2deg", Insight.rad2deg_fn(r).ptr != C_NULL)

println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0; exit(1); end
